### ISSUE TITLE
Error in prosody extraction in `03_gen_gan.sh`

### ISSUE THREAD

**Issue by unilight** [DigitalPhonetics/VoicePAT#12]
So I am executing `03_gen_gan.sh` and got the following error: (I think this happened when extracting vctk_dev_enrolls)
```
...
 25%|██▌       | 150/600 [00:56<02:20,  3.19it/s]
 25%|██▌       | 151/600 [00:56<02:25,  3.09it/s]
 25%|██▌       | 152/600 [00:57<02:29,  2.99it/s]
 26%|██▌       | 153/600 [00:57<02:37,  2.84it/s]
 26%|██▌       | 153/600 [00:59<02:54,  2.56it/s]
Traceback (most recent call last):
  File "run_anonymization.py", line 37, in <module>
    pipeline.run_anonymization_pipeline(datasets)
  File "/data/group1/z44476r/Experiments/VoicePAT/anonymization/pipelines/sttts_pipeline.py", line 130, in run_anonymization_pipeline
    prosody = self.prosody_extraction.extract_prosody(dataset_path=dataset_path, dataset_name=dataset_name,
  File "/data/group1/z44476r/Experiments/VoicePAT/anonymization/modules/prosody/prosody_extraction.py", line 58, in extract_prosody
    utt_prosody = self.extractor.extract_prosody(transcript=text, ref_audio_path=wav_path,
  File "/data/group1/z44476r/Experiments/VoicePAT/anonymization/modules/prosody/extraction/ims_prosody_extraction.py", line 99, in extract_prosody
    torch.nn.utils.clip_grad_norm_(acoustic_model.parameters(), 1.0)
  File "/home/z44476r/data/Experiments/VoicePAT/venv/lib/python3.8/site-packages/torch/nn/utils/clip_grad.py", line 76, in clip_grad_norm_
    torch._foreach_mul_(grads, clip_coef_clamped.to(device))  # type: ignore[call-overload]
RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```
In my experience, this is usually due to GPU OOM. I am using an NVIDIA V100, which has 32 GB and should be large enough.
So I went to check what happened in this line: https://github.com/DigitalPhonetics/VoicePAT/blob/vpc/anonymization/modules/prosody/extraction/ims_prosody_extraction.py#L99
It seems like some gradient calculation and model updating code. (I did not quite understand at first, but according to this comment I guess it's needed...? https://github.com/DigitalPhonetics/VoicePAT/blob/vpc/anonymization/modules/prosody/extraction/ims_prosody_extraction.py#L75)
Can I prevent this from happening?

**Comment by egaznep** [comment]
Could you post your environment specs? I did not have this error while testing on systems with 3090 Ti and 4090 Ti.

e.g., the output of `conda list`

**Comment by unilight** [comment]
```
# packages in environment at /home/z44476r/data/Experiments/VoicePAT/venv:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
_openmp_mutex             4.5                       1_gnu  
absl-py                   2.1.0                    pypi_0    pypi
annotated-types           0.6.0                    pypi_0    pypi
antlr4-python3-runtime    4.9.3                    pypi_0    pypi
asteroid-filterbanks      0.4.0                    pypi_0    pypi
attrs                     23.2.0                   pypi_0    pypi
audioread                 3.0.1                    pypi_0    pypi
babel                     2.14.0                   pypi_0    pypi
bibtexparser              2.0.0b6                  pypi_0    pypi
blas                      1.0                    openblas  
brotlipy                  0.7.0           py38h27cfd23_1003  
ca-certificates           2023.11.17           hbcca054_0    conda-forge
cachetools                5.3.2                    pypi_0    pypi
certifi                   2023.11.17         pyhd8ed1ab_0    conda-forge
cffi                      1.14.6           py38h400218f_0  
chardet                   4.0.0           py38h06a4308_1003  
ci-sdr                    0.0.2                    pypi_0    pypi
click                     8.1.7                    pypi_0    pypi
clldutils                 3.22.1                   pypi_0    pypi
cmake                     3.28.1                   pypi_0    pypi
colorama                  0.4.6                    pypi_0    pypi
colorlog                  6.8.2                    pypi_0    pypi
conda                     4.14.0           py38h578d9bd_0    conda-forge
conda-package-handling    1.7.3            py38h27cfd23_1  
configargparse            1.7                      pypi_0    pypi
cryptography              3.4.7            py38hd23ed53_0  
csvw                      3.3.0                    pypi_0    pypi
ctc-segmentation          1.7.4                    pypi_0    pypi
cvxopt                    1.3.0            py38h7aab8bc_0  
cycler                    0.12.1                   pypi_0    pypi
cython                    3.0.8                    pypi_0    pypi
decorator                 5.1.1                    pypi_0    pypi
distance                  0.1.3                    pypi_0    pypi
dlinfo                    1.2.1                    pypi_0    pypi
dsdp                      5.8               hd9d9efa_1203    conda-forge
editdistance              0.6.2                    pypi_0    pypi
einops                    0.7.0                    pypi_0    pypi
espnet                    202310                   pypi_0    pypi
espnet-model-zoo          0.1.7                    pypi_0    pypi
espnet-tts-frontend       0.0.3                    pypi_0    pypi
fast-bss-eval             0.1.3                    pypi_0    pypi
ffmpeg                    1.4                      pypi_0    pypi
fftw                      3.3.10          nompi_h77c792f_102    conda-forge
filelock                  3.13.1                   pypi_0    pypi
fsspec                    2023.12.2                pypi_0    pypi
future                    0.18.3                   pypi_0    pypi
g2p-en                    2.1.0                    pypi_0    pypi
glpk                      4.65              he80fd80_1002    conda-forge
gmp                       6.1.2             hf484d3e_1000    conda-forge
google-auth               2.27.0                   pypi_0    pypi
google-auth-oauthlib      1.0.0                    pypi_0    pypi
grpcio                    1.60.0                   pypi_0    pypi
gsl                       2.7.1                h6e86dc7_1  
h5py                      3.10.0                   pypi_0    pypi
huggingface-hub           0.20.3                   pypi_0    pypi
humanfriendly             10.0                     pypi_0    pypi
hydra-core                1.3.2                    pypi_0    pypi
hyperpyyaml               1.2.2                    pypi_0    pypi
idna                      2.10               pyhd3eb1b0_0  
importlib-metadata        4.13.0                   pypi_0    pypi
importlib-resources       6.1.1                    pypi_0    pypi
inflect                   7.0.0                    pypi_0    pypi
isodate                   0.6.1                    pypi_0    pypi
jaconv                    0.3.4                    pypi_0    pypi
jamo                      0.4.1                    pypi_0    pypi
jinja2                    3.1.3                    pypi_0    pypi
joblib                    1.3.2                    pypi_0    pypi
jsonschema                4.21.1                   pypi_0    pypi
jsonschema-specifications 2023.12.1                pypi_0    pypi
kaldiio                   2.18.0                   pypi_0    pypi
kiwisolver                1.4.5                    pypi_0    pypi
language-tags             1.2.0                    pypi_0    pypi
ld_impl_linux-64          2.35.1               h7274673_9  
libblas                   3.9.0           15_linux64_openblas    conda-forge
libcblas                  3.9.0           15_linux64_openblas    conda-forge
libffi                    3.3                  he6710b0_2  
libflac                   1.3.1                         0    conda-forge
libgcc-ng                 11.2.0               h1234567_1  
libgfortran-ng            13.2.0               h69a702a_0    conda-forge
libgfortran5              13.2.0               ha4646dd_0    conda-forge
libgomp                   11.2.0               h1234567_1  
liblapack                 3.9.0           15_linux64_openblas    conda-forge
libogg                    1.3.4                h7f98852_1    conda-forge
libopenblas               0.3.20          pthreads_h78a6416_0    conda-forge
librosa                   0.9.2                    pypi_0    pypi
libstdcxx-ng              9.3.0               hd4cf53a_17  
lit                       17.0.6                   pypi_0    pypi
llvmlite                  0.41.1                   pypi_0    pypi
lxml                      5.1.0                    pypi_0    pypi
markdown                  3.5.2                    pypi_0    pypi
markupsafe                2.1.4                    pypi_0    pypi
matplotlib                3.4.3                    pypi_0    pypi
metis                     5.1.0             h58526e2_1006    conda-forge
mpfr                      4.0.2                hb69a4c5_1  
mpmath                    1.3.0                    pypi_0    pypi
ncurses                   6.2                  he6710b0_1  
networkx                  3.1                      pypi_0    pypi
nltk                      3.8.1                    pypi_0    pypi
noisereduce               3.0.0                    pypi_0    pypi
numba                     0.58.0                   pypi_0    pypi
numpy                     1.23.5                   pypi_0    pypi
oauthlib                  3.2.2                    pypi_0    pypi
omegaconf                 2.3.0                    pypi_0    pypi
openssl                   1.1.1l               h7f98852_0    conda-forge
opt-einsum                3.3.0                    pypi_0    pypi
packaging                 23.2                     pypi_0    pypi
pandas                    2.0.3                    pypi_0    pypi
phonemizer                3.2.1                    pypi_0    pypi
pillow                    10.2.0                   pypi_0    pypi
pip                       21.1.3           py38h06a4308_0  
pkgutil-resolve-name      1.3.10                   pypi_0    pypi
platformdirs              4.1.0                    pypi_0    pypi
pooch                     1.8.0                    pypi_0    pypi
portaudio                 19.6.0               h7b6447c_4  
praat-parselmouth         0.4.3                    pypi_0    pypi
protobuf                  4.25.2                   pypi_0    pypi
pyasn1                    0.5.1                    pypi_0    pypi
pyasn1-modules            0.3.0                    pypi_0    pypi
pycosat                   0.6.3            py38h7b6447c_1  
pycparser                 2.20                       py_2  
pydantic                  2.6.0                    pypi_0    pypi
pydantic-core             2.16.1                   pypi_0    pypi
pylatexenc                2.10                     pypi_0    pypi
pyloudnorm                0.1.1                    pypi_0    pypi
pyopenssl                 20.0.1             pyhd3eb1b0_1  
pyparsing                 3.1.1                    pypi_0    pypi
pypinyin                  0.44.0                   pypi_0    pypi
pysocks                   1.7.1            py38h06a4308_0  
python                    3.8.10               h12debd9_8  
python-dateutil           2.8.2                    pypi_0    pypi
python-sounddevice        0.4.6              pyhd8ed1ab_0    conda-forge
python_abi                3.8                      2_cp38    conda-forge
pytorch-wpe               0.0.1                    pypi_0    pypi
pytz                      2023.4                   pypi_0    pypi
pyworld                   0.3.4                    pypi_0    pypi
pyyaml                    6.0.1                    pypi_0    pypi
rdflib                    7.0.0                    pypi_0    pypi
readline                  8.1                  h27cfd23_0  
referencing               0.33.0                   pypi_0    pypi
regex                     2023.12.25               pypi_0    pypi
requests                  2.25.1             pyhd3eb1b0_0  
requests-oauthlib         1.3.1                    pypi_0    pypi
resampy                   0.4.2                    pypi_0    pypi
rfc3986                   1.5.0                    pypi_0    pypi
rpds-py                   0.17.1                   pypi_0    pypi
rsa                       4.9                      pypi_0    pypi
ruamel-yaml               0.18.5                   pypi_0    pypi
ruamel-yaml-clib          0.2.8                    pypi_0    pypi
ruamel_yaml               0.15.100         py38h27cfd23_0  
scikit-learn              1.3.2                    pypi_0    pypi
scipy                     1.10.1                   pypi_0    pypi
segments                  2.2.1                    pypi_0    pypi
sentencepiece             0.1.97                   pypi_0    pypi
setuptools                52.0.0           py38h06a4308_0  
six                       1.16.0             pyhd3eb1b0_0  
soundfile                 0.12.1                   pypi_0    pypi
speechbrain               0.5.16                   pypi_0    pypi
sqlite                    3.36.0               hc218d9a_0  
suitesparse               5.10.1               hd8046ac_0    conda-forge
sympy                     1.12                     pypi_0    pypi
tabulate                  0.9.0                    pypi_0    pypi
tbb                       2020.2               h4bd325d_4    conda-forge
tensorboard               2.14.0                   pypi_0    pypi
tensorboard-data-server   0.7.2                    pypi_0    pypi
threadpoolctl             3.2.0                    pypi_0    pypi
tk                        8.6.10               hbc83047_0  
toolz                     0.12.1             pyhd8ed1ab_0    conda-forge
torch                     2.0.1+cu117              pypi_0    pypi
torch-complex             0.4.3                    pypi_0    pypi
torchaudio                2.0.2+cu117              pypi_0    pypi
torchvision               0.15.2+cu117             pypi_0    pypi
tqdm                      4.61.2             pyhd3eb1b0_1  
triton                    2.0.0                    pypi_0    pypi
typeguard                 2.13.3             pyhd8ed1ab_0    conda-forge
typing-extensions         4.9.0                    pypi_0    pypi
tzdata                    2023.4                   pypi_0    pypi
unidecode                 1.3.8                    pypi_0    pypi
uritemplate               4.1.1                    pypi_0    pypi
urllib3                   1.26.6             pyhd3eb1b0_1  
werkzeug                  3.0.1                    pypi_0    pypi
wheel                     0.36.2             pyhd3eb1b0_0  
xz                        5.2.5                h7b6447c_0  
yaml                      0.2.5                h7b6447c_0  
zipp                      3.17.0                   pypi_0    pypi
zlib                      1.2.11               h7b6447c_3  
```

**Comment by SarinaMeyer** [comment]
Hi,
The aligner for the prosody extraction is finetuned for each sample on the fly. It does improve the alignment, but you can also try it by disabling the finetuning (self.on_line_fine_tune=False) and see if the error persists.
Otherwise, I just recently saw that we produce quite a lot of overhead by always re-initializing the Aligner for each audio. You could try initializing the Aligner only in the init of ims_prosody_extraction.py and for each audio, you only re-set the initial weights (this is necessary to overwrite the audio-specific finetuning for a new audio).

Like this:

    def __init__(self, aligner_path, device, on_line_fine_tune=True, random_offset_lower=None,
                 random_offset_higher=None):
        ....
        self.acoustic_model = Aligner()
        self.aligner_weights = torch.load(aligner_path, map_location='cpu')["asr_model"]
        self.acoustic_model.load_state_dict(self.aligner_weights)
        self.acoustic_model = self.acoustic_model.to(self.device)

    def extract_prosody(self, transcript, ref_audio_path, lang="en", input_is_phones=False):
        # we need to reinitialize the aligner weights for each utterance again if we fine tune it for each utterance
        if self.on_line_fine_tune:
            self.acoustic_model.load_state_dict(self.aligner_weights)

        wave, sr = sf.read(ref_audio_path)
        ....

I hope that makes it work. 32 GB should definitely be more than enough.

**Comment by unilight** [comment]
@SarinaMeyer Thank you for the reply! (and sorry for getting back so late)

So the suggestion of moving the initialization code to `__init_` was not helping as the error that I encountered did not disappear. Instead, I just took your first advice and set `self.on_line_fine_tune=False`.
I'm still not sure why I got the original error. What's more confusing is that it seems to be somehow random. I ran the same command about 10 times and every time I get stuck at a different utterance.
